File: python/sigil_sdk/hooks.py
"""Synchronous hook evaluation for Sigil preflight/postflight rules."""

from __future__ import annotations

import base64
import json
from dataclasses import dataclass, field
from enum import Enum
from typing import Any
from urllib import error as urllib_error
from urllib import parse as urllib_parse
from urllib import request as urllib_request

from .config import HooksConfig
from .errors import HookDeniedError, HookTransportError
from .models import Message, PartKind, ToolDefinition

HOOKS_EVALUATE_PATH = "/api/v1/hooks:evaluate"
HOOK_TIMEOUT_HEADER = "X-Sigil-Hook-Timeout-Ms"
DEFAULT_HOOK_TIMEOUT = 15.0
_MAX_HOOK_RESPONSE_BYTES = 4 << 20


class HookPhase(str, Enum):
    """Hook evaluation phases."""

    PREFLIGHT = "preflight"
    POSTFLIGHT = "postflight"


class HookAction(str, Enum):
    """Verdict returned by the hook evaluation service."""

    ALLOW = "allow"
    DENY = "deny"


@dataclass(slots=True)
class HookModel:
    """Identifies the upstream model for hook rule matching."""

    provider: str = ""
    name: str = ""


@dataclass(slots=True)
class HookContext:
    """Routing/matching context attached to a hook evaluation request."""

    model: HookModel | None = None
    agent_name: str = ""
    agent_version: str = ""
    tags: dict[str, str] = field(default_factory=dict)


@dataclass(slots=True)
class HookInput:
    """Evaluable payload (request for preflight, request+response for postflight)."""

    messages: list[Message] = field(default_factory=list)
    tools: list[ToolDefinition] = field(default_factory=list)
    system_prompt: str = ""
    output: list[Message] = field(default_factory=list)
    conversation_preview: str = ""


@dataclass(slots=True)
class HookEvaluateRequest:
    """Hook evaluation request body."""

    phase: str
    context: HookContext
    input: HookInput


@dataclass(slots=True)
class HookEvaluation:
    """Per-rule outcome reported by the server."""

    rule_id: str
    evaluator_id: str
    evaluator_kind: str
    passed: bool
    latency_ms: int = 0
    explanation: str = ""
    reason: str = ""


@dataclass(slots=True)
class HookEvaluateResponse:
    """Hook evaluation response body."""

    action: str
    rule_id: str = ""
    reason: str = ""
    evaluations: list[HookEvaluation] = field(default_factory=list)

    @property
    def is_deny(self) -> bool:
        """Returns True when the server denied the request."""

        return self.action == HookAction.DENY.value


def hook_denied_from_response(response: HookEvaluateResponse | None) -> HookDeniedError | None:
    """Converts a denied evaluation response into a HookDeniedError."""

    if response is None or not response.is_deny:
        return None
    return HookDeniedError(
        reason=response.reason,
        rule_id=response.rule_id,
        evaluations=list(response.evaluations),
    )


def evaluate_hook(
    *,
    api_endpoint: str,
    insecure: bool,
    extra_headers: dict[str, str],
    hooks: HooksConfig,
    request: HookEvaluateRequest,
) -> HookEvaluateResponse:
    """Sends a hook evaluation request to the Sigil API.

    Returns ``HookAction.ALLOW`` without contacting the server when hooks are
    disabled or the request phase is not configured. Honours
    ``HooksConfig.fail_open`` to convert transport failures into allow
    responses (the default).
    """

    if not hooks.enabled:
        return _allow_response()

    phases = hooks.phases or ["preflight"]
    if request.phase not in phases:
        return _allow_response()

    timeout = hooks.timeout_seconds if hooks.timeout_seconds > 0 else DEFAULT_HOOK_TIMEOUT
    base_url = _base_url_from_api_endpoint(api_endpoint, insecure)
    if base_url is None:
        return _fail_open_or_raise(hooks.fail_open, "api endpoint is required")
    endpoint = base_url.rstrip("/") + HOOKS_EVALUATE_PATH

    timeout_ms = max(1, int(timeout * 1000))
    payload = json.dumps(_serialize_request(request)).encode("utf-8")
    headers = {
        "Content-Type": "application/json",
        HOOK_TIMEOUT_HEADER: str(timeout_ms),
        **(extra_headers or {}),
    }
    http_request = urllib_request.Request(
        endpoint,
        data=payload,
        method="POST",
        headers=headers,
    )

    try:
        with urllib_request.urlopen(http_request, timeout=timeout) as response:
            status = response.getcode()
            raw = response.read(_MAX_HOOK_RESPONSE_BYTES + 1)
    except urllib_error.HTTPError as exc:
        body = ""
        try:
            body = exc.read().decode("utf-8", errors="replace").strip()
        except Exception:  # noqa: BLE001
            body = ""
        message = body if body else f"HTTP {exc.code}"
        return _fail_open_or_raise(hooks.fail_open, f"status {exc.code}: {message}")
    except Exception as exc:  # noqa: BLE001
        return _fail_open_or_raise(hooks.fail_open, str(exc))

    if status < 200 or status >= 300:
        decoded = raw.decode("utf-8", errors="replace").strip()
        return _fail_open_or_raise(hooks.fail_open, f"status {status}: {decoded or 'unexpected status'}")
    if len(raw) > _MAX_HOOK_RESPONSE_BYTES:
        return _fail_open_or_raise(hooks.fail_open, "hook response too large")

    text = raw.decode("utf-8", errors="replace").strip()
    if text == "":
        return _fail_open_or_raise(hooks.fail_open, "empty hook response payload")

    try:
        parsed = json.loads(text)
    except Exception as exc:  # noqa: BLE001
        return _fail_open_or_raise(hooks.fail_open, f"invalid JSON response: {exc}")

    return _parse_response(parsed)


def _allow_response() -> HookEvaluateResponse:
    return HookEvaluateResponse(action=HookAction.ALLOW.value)


def _fail_open_or_raise(fail_open: bool, detail: str) -> HookEvaluateResponse:
    if fail_open:
        return _allow_response()
    raise HookTransportError(f"sigil hook evaluation failed: {detail}")


def _serialize_request(request: HookEvaluateRequest) -> dict[str, Any]:
    return {
        "phase": request.phase,
        "context": _serialize_context(request.context),
        "input": _serialize_input(request.input),
    }


def _serialize_context(context: HookContext) -> dict[str, Any]:
    out: dict[str, Any] = {}
    if context.model is not None:
        out["model"] = {
            "provider": context.model.provider,
            "name": context.model.name,
        }
    if context.agent_name:
        out["agent_name"] = context.agent_name
    if context.agent_version:
        out["agent_version"] = context.agent_version
    if context.tags:
        out["tags"] = dict(context.tags)
    return out


def _serialize_input(payload: HookInput) -> dict[str, Any]:
    out: dict[str, Any] = {}
    if payload.messages:
        out["messages"] = [_serialize_message(m) for m in payload.messages]
    if payload.tools:
        out["tools"] = [_serialize_tool(t) for t in payload.tools]
    if payload.system_prompt:
        out["system_prompt"] = payload.system_prompt
    if payload.output:
        out["output"] = [_serialize_message(m) for m in payload.output]
    if payload.conversation_preview:
        out["conversation_preview"] = payload.conversation_preview
    return out


def _message_role_wire(role: Any) -> int:
    """Maps SDK message roles to sigil.v1.MessageRole enum values."""

    value = getattr(role, "value", role)
    if value == "user":
        return 1
    if value == "assistant":
        return 2
    if value == "tool":
        return 3
    return 0


def _serialize_message(message: Message) -> dict[str, Any]:
    parts: list[dict[str, Any]] = []
    for part in message.parts:
        if part.kind == PartKind.TEXT and part.text:
            parts.append({"text": part.text})
        elif part.kind == PartKind.THINKING and part.thinking:
            parts.append({"thinking": part.thinking})
        elif part.kind == PartKind.TOOL_CALL and part.tool_call is not None:
            payload: dict[str, Any] = {
                "id": part.tool_call.id,
                "name": part.tool_call.name,
            }
            if part.tool_call.input_json:
                payload["input_json"] = base64.b64encode(part.tool_call.input_json).decode("ascii")
            parts.append({"tool_call": payload})
        elif part.kind == PartKind.TOOL_RESULT and part.tool_result is not None:
            tr = part.tool_result
            tr_payload: dict[str, Any] = {
                "is_error": tr.is_error,
                "content": tr.content,
            }
            if tr.tool_call_id:
                tr_payload["tool_call_id"] = tr.tool_call_id
            if tr.name:
                tr_payload["name"] = tr.name
            if tr.content_json:
                tr_payload["content_json"] = base64.b64encode(tr.content_json).decode("ascii")
            parts.append({"tool_result": tr_payload})
        else:
            # Fallback: emit a minimal text part so the payload remains valid JSON.
            if part.text:
                parts.append({"text": part.text})
    out: dict[str, Any] = {"role": _message_role_wire(message.role), "parts": parts}
    if message.name:
        out["name"] = message.name
    return out


def _serialize_tool(tool: ToolDefinition) -> dict[str, Any]:
    out: dict[str, Any] = {"name": tool.name}
    if tool.description:
        out["description"] = tool.description
    if tool.type:
        out["type"] = tool.type
    if tool.input_schema_json:
        out["input_schema_json"] = base64.b64encode(tool.input_schema_json).decode("ascii")
    if tool.deferred:
        out["deferred"] = True
    return out


def _parse_response(payload: Any) -> HookEvaluateResponse:
    if not isinstance(payload, dict):
        return _allow_response()
    action = payload.get("action")
    if action != HookAction.DENY.value:
        action = HookAction.ALLOW.value
    rule_id = _string_field(payload.get("rule_id"))
    reason = _string_field(payload.get("reason"))

    raw_evaluations = payload.get("evaluations")
    evaluations: list[HookEvaluation] = []
    if isinstance(raw_evaluations, list):
        for entry in raw_evaluations:
            if not isinstance(entry, dict):
                continue
            evaluations.append(
                HookEvaluation(
                    rule_id=_string_field(entry.get("rule_id")),
                    evaluator_id=_string_field(entry.get("evaluator_id")),
                    evaluator_kind=_string_field(entry.get("evaluator_kind")),
                    passed=bool(entry.get("passed")),
                    latency_ms=_int_field(entry.get("latency_ms")),
                    explanation=_string_field(entry.get("explanation")),
                    reason=_string_field(entry.get("reason")),
                )
            )
    return HookEvaluateResponse(
        action=action,
        rule_id=rule_id,
        reason=reason,
        evaluations=evaluations,
    )


def _string_field(value: Any) -> str:
    if isinstance(value, str):
        return value
    return ""


def _int_field(value: Any) -> int:
    if isinstance(value, bool):
        return 0
    if isinstance(value, int):
        return value
    if isinstance(value, float):
        try:
            return int(value)
        except (OverflowError, ValueError):
            return 0
    if isinstance(value, str):
        try:
            return int(value)
        except ValueError:
            return 0
    return 0


def _base_url_from_api_endpoint(endpoint: str, insecure: bool) -> str | None:
    trimmed = (endpoint or "").strip()
    if trimmed == "":
        return None
    if trimmed.startswith("http://") or trimmed.startswith("https://"):
        parsed = urllib_parse.urlparse(trimmed)
        if not parsed.scheme or not parsed.netloc:
            return None
        return f"{parsed.scheme}://{parsed.netloc}"
    without_scheme = trimmed[7:] if trimmed.startswith("grpc://") else trimmed
    host = without_scheme.split("/", 1)[0].strip()
    if host == "":
        return None
    scheme = "http" if insecure else "https"
    return f"{scheme}://{host}"

File: js/src/hooks.ts
import type {
  HookEvaluateRequest,
  HookEvaluateResponse,
  HookEvaluation,
  HookInput,
  HooksConfig,
} from './types.js';
import { asError } from './utils.js';

const hooksEvaluatePath = '/api/v1/hooks:evaluate';
const hookTimeoutHeader = 'X-Sigil-Hook-Timeout-Ms';

/**
 * Thrown by framework adapters when hook evaluation returns `action: 'deny'`.
 *
 * The error preserves the rule that triggered the deny and the per-rule
 * evaluation outcomes so callers can build user-facing error messages.
 */
export class HookDeniedError extends Error {
  readonly action = 'deny' as const;
  readonly reason: string;
  readonly ruleId?: string;
  readonly evaluations: HookEvaluation[];

  constructor(reason: string, ruleId?: string, evaluations: HookEvaluation[] = []) {
    super(formatDenyMessage(reason, ruleId));
    this.name = 'HookDeniedError';
    const normalized = reason?.trim() ?? '';
    this.reason = normalized.length > 0 ? normalized : 'request blocked by Sigil hook rule';
    this.ruleId = ruleId;
    this.evaluations = evaluations;
  }
}

/**
 * Sends a hook evaluation request to the Sigil API.
 *
 * `apiEndpoint` is the Sigil API base URL (without the `/api/v1/...` suffix).
 * `extraHeaders` is merged into the request — typically the same auth headers
 * the SDK uses for generation export.
 *
 * Returns `{ action: 'allow', evaluations: [] }` when the request fails and
 * `hooks.failOpen` is true.
 */
export async function evaluateHook(params: {
  apiEndpoint: string;
  insecure: boolean;
  extraHeaders: Record<string, string> | undefined;
  hooks: HooksConfig;
  request: HookEvaluateRequest;
  fetchImpl?: typeof fetch;
}): Promise<HookEvaluateResponse> {
  const fetchImpl = params.fetchImpl ?? fetch;
  if (!params.hooks.enabled) {
    return allowResponse();
  }

  const phases = params.hooks.phases;
  if (phases.length > 0 && !phases.includes(params.request.phase)) {
    return allowResponse();
  }

  const timeoutMs = params.hooks.timeoutMs > 0 ? params.hooks.timeoutMs : 15_000;
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeoutMs);

  try {
    const url = buildHooksEvaluateEndpoint(params.apiEndpoint, params.insecure);
    const body = serializeRequest(params.request);

    const response = await fetchImpl(url, {
      method: 'POST',
      signal: controller.signal,
      headers: {
        'content-type': 'application/json',
        [hookTimeoutHeader]: String(timeoutMs),
        ...(params.extraHeaders ?? {}),
      },
      body: JSON.stringify(body),
    });

    const responseText = (await response.text()).trim();
    if (!response.ok) {
      return failOpenOrThrow(
        params.hooks.failOpen,
        new Error(
          `sigil hook evaluation failed: status ${response.status}: ${hookErrorText(responseText, response.status)}`,
        ),
      );
    }
    if (responseText.length === 0) {
      return failOpenOrThrow(params.hooks.failOpen, new Error('sigil hook evaluation failed: empty response payload'));
    }

    let payload: unknown;
    try {
      payload = JSON.parse(responseText);
    } catch (error) {
      return failOpenOrThrow(
        params.hooks.failOpen,
        new Error(`sigil hook evaluation failed: invalid JSON response: ${asError(error).message}`),
      );
    }

    return parseEvaluateResponse(payload);
  } catch (error) {
    return failOpenOrThrow(params.hooks.failOpen, asError(error));
  } finally {
    clearTimeout(timer);
  }
}

function failOpenOrThrow(failOpen: boolean, error: Error): HookEvaluateResponse {
  if (failOpen) {
    return allowResponse();
  }
  throw error;
}

function allowResponse(): HookEvaluateResponse {
  return { action: 'allow', evaluations: [] };
}

function formatDenyMessage(reason: string, ruleId: string | undefined): string {
  const trimmedReason = reason?.trim() ?? '';
  const baseReason = trimmedReason.length > 0 ? trimmedReason : 'request blocked by Sigil hook rule';
  if (ruleId !== undefined && ruleId.length > 0) {
    return `sigil hook denied by rule ${ruleId}: ${baseReason}`;
  }
  return `sigil hook denied: ${baseReason}`;
}

function buildHooksEvaluateEndpoint(endpoint: string, insecure: boolean): string {
  const baseURL = baseURLFromAPIEndpoint(endpoint, insecure);
  return `${baseURL}${hooksEvaluatePath}`;
}

function baseURLFromAPIEndpoint(endpoint: string, insecure: boolean): string {
  const trimmed = endpoint.trim();
  if (trimmed.length === 0) {
    throw new Error('sigil hook evaluation failed: api endpoint is required');
  }

  if (trimmed.startsWith('http://') || trimmed.startsWith('https://')) {
    const parsed = new URL(trimmed);
    return `${parsed.protocol}//${parsed.host}`;
  }

  const withoutScheme = trimmed.startsWith('grpc://') ? trimmed.slice('grpc://'.length) : trimmed;
  const host = withoutScheme.split('/')[0]?.trim();
  if (host === undefined || host.length === 0) {
    throw new Error('sigil hook evaluation failed: api endpoint host is required');
  }
  return `${insecure ? 'http' : 'https'}://${host}`;
}

function serializeRequest(request: HookEvaluateRequest): Record<string, unknown> {
  const body: Record<string, unknown> = {
    phase: request.phase,
    context: serializeContext(request.context),
    input: serializeInput(request.input),
  };
  return body;
}

function serializeContext(context: HookEvaluateRequest['context']): Record<string, unknown> {
  const out: Record<string, unknown> = {
    model: { provider: context.model.provider, name: context.model.name },
  };
  if (context.agentName !== undefined && context.agentName.length > 0) {
    out.agent_name = context.agentName;
  }
  if (context.agentVersion !== undefined && context.agentVersion.length > 0) {
    out.agent_version = context.agentVersion;
  }
  if (context.tags !== undefined && Object.keys(context.tags).length > 0) {
    out.tags = { ...context.tags };
  }
  return out;
}

function serializeInput(input: HookEvaluateRequest['input']): Record<string, unknown> {
  const out: Record<string, unknown> = {};
  if (input.messages !== undefined && input.messages.length > 0) {
    out.messages = input.messages;
  }
  if (input.tools !== undefined && input.tools.length > 0) {
    out.tools = input.tools;
  }
  if (input.systemPrompt !== undefined && input.systemPrompt.length > 0) {
    out.system_prompt = input.systemPrompt;
  }
  if (input.output !== undefined && input.output.length > 0) {
    out.output = input.output;
  }
  if (input.conversationPreview !== undefined && input.conversationPreview.length > 0) {
    out.conversation_preview = input.conversationPreview;
  }
  return out;
}

function parseEvaluateResponse(payload: unknown): HookEvaluateResponse {
  if (!isRecord(payload)) {
    throw new Error('sigil hook evaluation failed: invalid response payload');
  }
  const action = payload.action === 'deny' ? 'deny' : 'allow';
  const ruleId = typeof payload.rule_id === 'string' ? payload.rule_id : undefined;
  const reason = typeof payload.reason === 'string' ? payload.reason : undefined;

  const rawEvaluations = Array.isArray(payload.evaluations) ? payload.evaluations : [];
  const evaluations: HookEvaluation[] = [];
  for (const entry of rawEvaluations) {
    if (!isRecord(entry)) {
      continue;
    }
    evaluations.push({
      ruleId: stringField(entry.rule_id),
      evaluatorId: stringField(entry.evaluator_id),
      evaluatorKind: stringField(entry.evaluator_kind),
      passed: entry.passed === true,
      latencyMs: numberField(entry.latency_ms),
      explanation: optionalStringField(entry.explanation),
      reason: optionalStringField(entry.reason),
    });
  }

  const transformedInput = parseTransformedInputPayload(payload);

  return { action, ruleId, reason, transformedInput, evaluations };
}

function parseTransformedInputPayload(payload: Record<string, unknown>): HookInput | undefined {
  const raw = payload.transformed_input;
  if (raw === undefined || raw === null) {
    return undefined;
  }
  if (!isRecord(raw)) {
    return undefined;
  }
  const out: HookInput = {};
  if (Array.isArray(raw.messages) && raw.messages.length > 0) {
    out.messages = raw.messages as HookInput['messages'];
  }
  if (Array.isArray(raw.tools) && raw.tools.length > 0) {
    out.tools = raw.tools as HookInput['tools'];
  }
  if (typeof raw.system_prompt === 'string' && raw.system_prompt.length > 0) {
    out.systemPrompt = raw.system_prompt;
  }
  if (Array.isArray(raw.output) && raw.output.length > 0) {
    out.output = raw.output as HookInput['output'];
  }
  if (typeof raw.conversation_preview === 'string' && raw.conversation_preview.length > 0) {
    out.conversationPreview = raw.conversation_preview;
  }
  if (
    out.messages === undefined &&
    out.tools === undefined &&
    out.systemPrompt === undefined &&
    out.output === undefined &&
    out.conversationPreview === undefined
  ) {
    return undefined;
  }
  return out;
}

function stringField(value: unknown): string {
  return typeof value === 'string' ? value : '';
}

function optionalStringField(value: unknown): string | undefined {
  if (typeof value !== 'string' || value.length === 0) {
    return undefined;
  }
  return value;
}

function numberField(value: unknown): number {
  if (typeof value === 'number' && Number.isFinite(value)) {
    return value;
  }
  return 0;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

function hookErrorText(body: string, status: number): string {
  if (body.length > 0) {
    return body;
  }
  return `HTTP ${status}`;
}
